### ISSUE TITLE
Constraint files

### ISSUE THREAD

**Issue by IvanAnishchuk** [peterdemin/pip-compile-multi#91]
It seems that this tool doesn't understand constraint files (pip `-c` flag) at all. They are just like referenced requirements files except they own limit your requirements versions and don't add any new packages to install.

What I want to get in the end: prod.in has `-c base.in` and test.in has `-c base.in` and `-c prod.in` lines in them. When I compile, I want versions in all three resulting files (base.txt, prod.txt, test.txt) to be synced and `-c` lines replaced with compiled files (so prod.txt has `-c base.txt` and test.txt `-c base.txt` and `-c prod.txt`). This way I can select to install everything or to install only test packages or only prod ones (or split them into different docker layers), for that it's very important that test reqs file doesn't include requirements from prod.txt unless there's some dependency but versions of those sub-deps are synced nevetherless.

It also would be great to have a dependency-based compilation order if possible.

Does this sound like something worth implementing in this tool?

**Comment by IvanAnishchuk** [comment]
Actually, come to think of it. Referencing constraints in compiled files is not necessary as long as constraints never change without complete recompilation... Skipping them in pinned files is a pip-tools behaviour and it actually does respect the constraints as far as I can tell.

Still, dependency-based order of compilation is definitely needed.

**Comment by peterdemin** [comment]
If `.in` files use other `.in` files for constraints and `.txt` files don't have `-c`, than compilation order doesn't matter.
I used `conflicting-in-merge` example for experiments. Without `-c` files are:
```
==> conflicting-in-merge/base1.in <==
pytz<2018

==> conflicting-in-merge/base2.in <==
pytz

==> conflicting-in-merge/together.in <==
-r base1.in
-r base2.in
```
Compilation gives an error:
```
$ pip-compile-multi -d conflicting-in-merge/
Locking conflicting-in-merge/base1.in to conflicting-in-merge/base1.txt. References: []
Locking conflicting-in-merge/base2.in to conflicting-in-merge/base2.txt. References: []
Package pytz was resolved to different versions in different environments: 2018.7 and 2017.3
```
But if I add `-c base1.in` to `base2.in`:
```
==> conflicting-in-merge/base1.in <==
pytz<2018

==> conflicting-in-merge/base2.in <==
-c base1.in

pytz

==> conflicting-in-merge/together.in <==
-r base1.in
-r base2.in
```
Compilation succeeds with expected result:
```
$ head conflicting-in-merge/*.txt
==> conflicting-in-merge/base1.txt <==
# SHA1:07973b0a74462beb75c2bb65cae3db0b53237103
#
# This file is autogenerated by pip-compile-multi
# To update, run:
#
#    pip-compile-multi
#
pytz==2017.3

==> conflicting-in-merge/base2.txt <==
# SHA1:d3d3929908415fe98d4376a9794073ccc3f61c8f
#
# This file is autogenerated by pip-compile-multi
# To update, run:
#
#    pip-compile-multi
#
pytz==2017.3

==> conflicting-in-merge/together.txt <==
# SHA1:08d9c33c7a4a6c4972d515c176318068baf08856
#
# This file is autogenerated by pip-compile-multi
# To update, run:
#
#    pip-compile-multi
#
-r base1.txt
-r base2.txt
```

I realize, that it may be hard to follow the example. And maybe you have a different case where constraint files do not work as expected. So please share your requirements files and I'll add them to the test suite.

**Comment by peterdemin** [comment]
I'm closing it for now, feel free to reopen.
